src/indexer/json_term_writer.rs

```diff
@@ -199,12 +199,74 @@ fn infer_type_from_str(text: &str) -> TextOrDateTime {
     }
 }
 
+// Tries to infer a JSON type from a string
+pub(crate) fn convert_to_fast_value_and_get_term(
+    json_term_writer: &mut JsonTermWriter,
+    phrase: &str,
+) -> Option<Term> {
+    if let Ok(dt) = OffsetDateTime::parse(phrase, &Rfc3339) {
+        let dt_utc = dt.to_offset(UtcOffset::UTC);
+        return Some(set_fastvalue_and_get_term(
+            json_term_writer,
+            DateTime::from_utc(dt_utc),
+        ));
+    }
+    if let Ok(u64_val) = str::parse::<u64>(phrase) {
+        return Some(set_fastvalue_and_get_term(json_term_writer, u64_val));
+    }
+    if let Ok(i64_val) = str::parse::<i64>(phrase) {
+        return Some(set_fastvalue_and_get_term(json_term_writer, i64_val));
+    }
+    if let Ok(f64_val) = str::parse::<f64>(phrase) {
+        return Some(set_fastvalue_and_get_term(json_term_writer, f64_val));
+    }
+    None
+}
+// helper function to generate a Term from a json fastvalue
+pub(crate) fn set_fastvalue_and_get_term<T: FastValue>(
+    json_term_writer: &mut JsonTermWriter,
+    value: T,
+) -> Term {
+    json_term_writer.set_fast_value(value);
+    json_term_writer.term().clone()
+}
+
+// helper function to generate a list of terms with their positions from a textual json value
+pub(crate) fn set_string_and_get_terms(
+    json_term_writer: &mut JsonTermWriter,
+    value: &str,
+    text_analyzer: &TextAnalyzer,
+) -> Vec<(usize, Term)> {
+    let mut positions_and_terms = Vec::<(usize, Term)>::new();
+    json_term_writer.close_path_and_set_type(Type::Str);
+    let term_num_bytes = json_term_writer.term_buffer.as_slice().len();
+    let mut token_stream = text_analyzer.token_stream(value);
+    token_stream.process(&mut |token| {
+        json_term_writer.term_buffer.truncate(term_num_bytes);
+        json_term_writer
+            .term_buffer
+            .append_bytes(token.text.as_bytes());
+        positions_and_terms.push((token.position, json_term_writer.term().clone()));
+    });
+    positions_and_terms
+}
+
 pub struct JsonTermWriter<'a> {
     term_buffer: &'a mut Term,
     path_stack: Vec<usize>,
 }
 
 impl<'a> JsonTermWriter<'a> {
+    // Prepares writing terms for a given field
+    pub fn initialize(field: Field, json_path: &str, term_buffer: &'a mut Term) -> Self {
```

Comment:
initialize is quite uncommon for a constructor, typical is `new` or `from_field_and_json_path`.

Can we move ownership of the Term into the JsonTermWriter? Currently it's unclear why it's passed as mut ref (performance or are the mutated object contents used somewhere else?)

Reply:
@PSeitz I'll rename it to `from_field_and_json_path`, maybe we also need to rename `wrap` ? 

It looks like we pass the mut ref for performance cf https://github.com/saroh/tantivy/blob/706fbd68868523c4e41ff3341f8a94f37f9f44f2/src/indexer/segment_writer.rs#L276-L288 , should I still move the ownership ?

Reply:
> Maybe it could make sense to make JsonTermWriter private for the crate (unsure about the scope here), and expose another struct for general use which would simply wrap a JsonTermWriter with ownership of the term buffer ?

Yeah it's similar to `Term`, which has some API weirdness, which may be improved when splitting in private and public.

> Also do we need to make sure the json_path is not empty string or is this to be validated elsewhere/ OK for the time being?

I don't see `path()` anywhere used in tantivy. If you see  a place where you can add an assertion feel free to add one.

Reply:
After a bit of searching, the specs states we should be OK as is and the check is unnecessary.
![image](https://user-images.githubusercontent.com/325288/167146715-000ab520-b779-4140-a17e-cbc0d411cb1a.png)
https://www.ecma-international.org/publications-and-standards/standards/ecma-404/